// @paramkhodiyar/chat-core/dist/rag.js
export class RagManager {
    chunks = [];
    enabled;
    constructor(enabled = false) {
        this.enabled = enabled;
    }
    // Very naive in-memory "semantic" search for V1
    // Real implementation would use embeddings and cosine similarity
    async search(query, topK = 3) {
        if (!this.enabled || this.chunks.length === 0)
            return [];
        const queryTokens = query.toLowerCase().split(/\s+/);
        // Sort by simple keyword overlap
        const scored = this.chunks.map(chunk => {
            let score = 0;
            const text = chunk.text.toLowerCase();
            for (const token of queryTokens) {
                if (text.includes(token))
                    score++;
            }
            return { chunk, score };
        });
        scored.sort((a, b) => b.score - a.score);
        return scored.slice(0, topK).map(s => s.chunk);
    }
    ingestDocument(text, metadata) {
        // Basic chunking by paragraphs
        const paragraphs = text.split('\n\n');
        paragraphs.forEach((p, idx) => {
            if (p.trim().length > 10) {
                this.chunks.push({
                    id: `chunk-${Date.now()}-${idx}`,
                    text: p.trim(),
                    metadata
                });
            }
        });
    }
}
